cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -160,6 +160,9 @@ func Run(s *options.CMServer) error {
 	if err != nil {
 		return err
 	}
+	// it is more safe to use hostname plus pod name
+	// see https://github.com/kubernetes/kubernetes/issues/22775
+	id = id + "-" + os.Getenv("POD_NAME")
```

Comment:
What you mentioned is absolutely correct.

Let's do what you suggest, once the scheduler is running as a real pod via replica set without code change, I have two concerns here still:

1.  How could we ensure there is no port conflict(if they are running on the same node).
2.  Considering the same situation, we could not do the log volume mount, either.